src/coprocessor/codec/mysql/decimal.rs

```diff
@@ -1428,6 +1466,35 @@ impl Decimal {
         Res::Ok(x)
     }
 
+    /// `int_part` returns int part of the decimal. It's temporary and
+    /// after we adjust `as_i64`, it will be removed.
+    pub fn int_part(&self, ctx: &EvalContext) -> ::std::result::Result<i64, ExprError> {
```

Comment:
`as_i64` doesn't retrieve `StatementContext` argument, so that if any `Truncate` occurs, it can only return it by `Res::Truncate`. It causes that we need to test that everywhere (e.g. `ceil/floor` and `cast`). We need to write code everywhere like this:
```rust
fn ceil_dec_to_int(...) {
    match dec.as_i64() {
        Res::Ok(t) => t,
        Truncate(t) => None, // something here.
        _ => None, // some other thing.
    }
}
```

So I think we can pass `Statement` into `as_i64`. If any `truncate` occurs inside that function, we just need to check the flag, and return `Ok(t)` or `Err` directly.